packages/xstate-solid/src/useMachine.ts

```diff
@@ -0,0 +1,37 @@
+import type { AnyStateMachine, StateFrom } from 'xstate';
+import type { CheckSnapshot, RestParams } from './types';
+import { createService } from './createService';
+import { createRenderEffect, onCleanup } from 'solid-js';
+import type { InterpreterFrom, Prop } from 'xstate';
+import { deriveServiceState } from './deriveServiceState';
+import { createImmutable } from './createImmutable';
+
+type UseMachineReturn<
+  TMachine extends AnyStateMachine,
+  TInterpreter = InterpreterFrom<TMachine>
+> = [
+  CheckSnapshot<StateFrom<TMachine>>,
+  Prop<TInterpreter, 'send'>,
+  TInterpreter
+];
+
+export function useMachine<TMachine extends AnyStateMachine>(
+  machine: TMachine,
+  ...[options = {}]: RestParams<TMachine>
+): UseMachineReturn<TMachine> {
+  const service = createService(machine, options);
+
+  const [state, setState] = createImmutable(
+    deriveServiceState(service, service.state)
+  );
+
+  createRenderEffect(() => {
```

Comment:
since we don't deal with changing sources here - I would probably use `createEffect` here

Reply:
I'd like to keep this a `createRenderEffect` so synchronous updates can happen in the setup body of the component for both CSR and SSR